### PR TITLE
DrivAerML: Lion optimizer with aggressive LRs (5e-4, 3e-4, 1e-3)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
Lion optimizer was used successfully for TandemFoil but has not been tested on DrivAerML. Lion uses sign-based updates (momentum), which can be more aggressive than AdamW and sometimes converges faster or to better minima on geometry tasks. The ~0.1x LR rule of thumb for Lion vs AdamW would suggest lr ~5e-5 for DrivAerML (since AdamW uses 5e-4), but Lion often tolerates higher LRs in practice. We test Lion at three scales: 5e-4 (same absolute value as AdamW baseline for direct comparison), 3e-4 (intermediate), and 1e-3 (aggressive, capitalizing on Lion's momentum-like property).

## Instructions
Test Lion optimizer at three learning rates on DrivAerML 4L/512d architecture. Keep gc=1.0, WD=1e-2, T_max=30.

**Run 1 (CUDA 0) — Lion lr=5e-4:**
```bash
CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset drivaerml \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --optimizer lion \
  --lr 5e-4 \
  --weight-decay 1e-2 \
  --grad-clip 1.0 \
  --cosine-t-max 30 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --wandb-group drivaerml-lion \
  --wandb-name driv-lion-lr5e4
```

**Run 2 (CUDA 1) — Lion lr=3e-4:**
```bash
CUDA_VISIBLE_DEVICES=1 python train.py \
  --dataset drivaerml \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --optimizer lion \
  --lr 3e-4 \
  --weight-decay 1e-2 \
  --grad-clip 1.0 \
  --cosine-t-max 30 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --wandb-group drivaerml-lion \
  --wandb-name driv-lion-lr3e4
```

**Run 3 (CUDA 2) — Lion lr=1e-3 (aggressive):**
```bash
CUDA_VISIBLE_DEVICES=2 python train.py \
  --dataset drivaerml \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --optimizer lion \
  --lr 1e-3 \
  --weight-decay 1e-2 \
  --grad-clip 1.0 \
  --cosine-t-max 30 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --wandb-group drivaerml-lion \
  --wandb-name driv-lion-lr1e3
```

Note: If any run diverges (loss spikes or NaN), terminate early and report the last stable loss value. Lion can be unstable at high LRs — the 1e-3 run is deliberately aggressive to find the edge.

## Baseline
- Dataset: DrivAerML
- Primary metric: percentage error (lower is better)
- Current best: **4.619%** (PR #2691, 4L/512d/8H + AdamW lr=5e-4 + T_max=30)

## Expected Outcome
Report percentage error for each run. If any Lion variant beats 4.619%, we have a new optimizer direction for DrivAerML. Compare Lion lr=5e-4 vs AdamW lr=5e-4 directly to isolate the optimizer effect. If Lion is unstable at 1e-3, the optimal Lion LR is likely between 3e-4 and 5e-4.

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT giyu:

## Results

### Summary — All 3 Lion runs diverged on DrivAerML

| Run | LR | Best val_rel_l2 | Best Epoch | Diverged Epoch | Terminal grad_norm | W&B |
|---|---|---|---|---|---|---|
| 1 | 5e-4 | **23.70%** | ep15 | ep16 | 43.5M (ep18) | mppk4eu8 |
| 2 | 3e-4 | 30.37% | ep10 | ep11 | 881K (ep15) | gkyec2ru |
| 3 | 1e-3 | 62.79% | ep1 | ep3 | 4.2B (ep4) | 4ebhju4o |

**Baseline: 4.619%** (AdamW lr=5e-4, PR #2691)

None of the Lion runs came close to the baseline. All diverged catastrophically.

### Divergence timeline

**lr=5e-4** (longest survivor):
- Excellent early convergence: trough envelope 37.07→31.56→29.10→26.41→23.70%
- ep16: sudden gradient explosion (grad_norm=38,708, val=75.10%)
- ep18: terminal (grad_norm=43.5M, val=171.6%)

**lr=3e-4:**
- Trough envelope: 38.91→33.76→30.37%
- ep11: instability begins (val=48.22%, grad_norm=4.49)
- ep12: gradient explosion (grad_norm=558, val=60.39%)
- ep15: terminal (grad_norm=881K, val=75.15%)

**lr=1e-3:**
- ep1: 62.79% (only stable epoch)
- ep3: gradient explosion (grad_norm=35,509)
- ep4: terminal (grad_norm=4.2B, val=389%). Killed immediately.

### Key question answered: Does Lion beat AdamW on DrivAerML?

**No.** Lion is fundamentally incompatible with DrivAerML's training regime. Lower LR delays divergence but does not prevent it. The pattern holds across all 3 LRs: excellent early convergence → sudden catastrophic gradient explosion → terminal model death.

### What happened — analysis

Lion's sign-based updates are too aggressive for DrivAerML's extreme conditions:
1. **batch_size=1** means gradient estimates are very noisy
2. Sign-based updates amplify noise (sign(noisy_grad) ≈ random ±1 for small gradients)
3. At cosine LR peaks, the sign-based steps become large enough to destabilize
4. Once destabilized, gradient clipping (gc=1.0) cannot recover the model — the exploding gradient norm (38K→43.5M in 2 epochs) overwhelms the clip

Interestingly, the "0.1x LR rule" for Lion vs AdamW did NOT hold: lr=5e-4 (same as AdamW baseline) was actually the best Lion LR, surviving 15 epochs vs lr=3e-4's 10 epochs. This may be because the lower LR gives Lion even more step-to-noise ratio issues with batch_size=1.

### Commands used

Note: Added `--no-use-ema --enable-fourier --epochs 999` to match baseline config (missing from PR).

```bash
# lr=5e-4 (GPU 0)
CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset drivaerml --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --optimizer lion --lr 5e-4 --weight-decay 1e-2 --grad-clip 1.0 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 --epochs 999 \
  --agent giyu --wandb-group drivaerml-lion --wandb-name "giyu/driv-lion-lr5e4"

# lr=3e-4 (GPU 1) — same but --lr 3e-4
# lr=1e-3 (GPU 2) — same but --lr 1e-3
```

### Suggested follow-ups

1. **Lion + higher gc (gc=0.1 or gc=0.5)** — tighter clipping might prevent the gradient explosion at cosine peaks, at the cost of slower convergence
2. **Lion + larger batch size** — if DrivAerML supports batch_size>1 without OOM, this would reduce gradient noise and make sign-based updates more stable
3. **Lion + longer T_max (T_max=60 or T_max=100)** — less aggressive LR cycling might avoid the instability trigger at cosine peaks
4. **Lion on TandemFoil/AirfRANS at 4L/512d** — Lion already works on TandemFoil at 3L/192d; worth testing if it transfers to DrivAerML-scale architectures on other datasets

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2871 because ADVISOR: Closing — Lion optimizer is a dead end on DrivAerML. All three LRs failed catastrophically (best: 26.41% at lr=3e-4, 5.7x worse than 4.619% baseline). The divergence is structural: Lion's sign-based updates + batch_size=1 noisy gradients + cosine LR peaks = uncontrollable gradient explosions. Your analysis was thorough and accurate — important negative finding that rules out Lion for 3D CFD surrogates at this batch size. You'll be reassigned to AirfRANS where the 2L depth frontier is hot..
